File: iOS/Puzzlr/Hyperledger/PublicKeyBlockchain.swift
//
//  PublicKeyBlockchain.swift
//  Puzzlr
//
//  This class represent the Blockchain containing user's public key.
//
//  Created by Quentin Le Sceller on 09/03/2016.
//  Copyright © 2016 Quentin Le Sceller. All rights reserved.
//

import Foundation

class PublicKeyBlockchain {
    
    // The main blokchain:
    let blockchain : OpenBlockchain
    
    // The user blockchain:
    let userBlockchain : UserBlockchain
    
    // The language used for the chaincode.
    let type : String
    
    // The name of the chaincode (usually hash value).
    let name : String
    
    // Initializer
    init(ip: String,port : Int) {
        blockchain = OpenBlockchain(ip: ip, port: port)
        type = "GOLANG";
        name = "7e60d53a09aff0a72728b64decb0600ceae25339ca5389e6b3c5d48116a00df3c753c06da58b4c906bd6cd71d1f91827938b50f468be43b27bb1dc700bd0df7f";
        userBlockchain = UserBlockchain(ip: ip, port: port)
    }
    
    /**
     Register public key on blockchain.
     
     - parameter user:      the username
     - parameter publickey: user's publickey
     - parameter sync:      synchronous or asynchronous request
     
     - returns: true if successfully registered, false otherwise:
     */
    func registerPublicKey(user : String, publickey : String, sync : Bool) -> Bool {
        var success = Bool(true)
        let args = [user, publickey]
        
        if userBlockchain.getRegistered(user, sync: sync) {
            blockchain.invoke(type, name: name, function: "invoke", args: args, sync: sync)
        } else {
            success = Bool(false)
            print("User is not registered")
        }
        print("success!")
        return success
    }
    
    /**
     Delete public key on blockchain.
     
     - parameter user:      the username
     - parameter publickey: user's publickey
     - parameter sync:      synchronous or asynchronous request
     
     - returns: true if successfully deleted, false otherwise:
     */
    func deletePublicKey(user : String, password : String, sync : Bool)-> Bool {
        var success = Bool(true)
        let args = [user]
        
        if (userBlockchain.login(user, password: password, sync: sync)){
            blockchain.invoke(type, name: name, function: "delete", args: args, sync: sync)
        } else {
            success = Bool(false)
        }
        return success
    }
    
    /**
     Query public key linked to an username.
     
     - parameter user: the username
     - parameter sync: synchronous or asynchronous request
     
     - returns:  true if successfully retrieved, false otherwise:
     */
    func queryPublicKey(user : String, sync : Bool) -> NSString {
        var publickey : NSString?
        let args = [user]
        
        if userBlockchain.getRegistered(user, sync: sync){
            let result = blockchain.query(type, name: name, function: "query", args: args, sync: sync)
            publickey = result.substringFromIndex(7)
            let index1 = publickey?.length.advancedBy(-2)
            publickey = publickey?.substringToIndex(index1!)
        } else {
            publickey = "null"
        }
        
        return publickey!
    }
    
    
}

File: Android/Puzzlr/app/src/main/java/BlockChain/PublicKeyBlockchain.java
package BlockChain;

/**
 * Created by aniss on 09/03/16.
 */
public class PublicKeyBlockchain implements IPublicKeyBlockchain {

    /** The blockchain. */
    OpenBlockchain blockchain;

    /** The user blockchain. */
    UserBlockchain userBlockchain;

    /** The type. */
    String type;

    /** The name. */
    String name;

    /**
     * Instantiates a new public key blockchain.
     *
     * @param address
     *            the address
     * @param port
     *            the port
     */
    public PublicKeyBlockchain(String address, int port) {
        blockchain = new OpenBlockchain(address, port);

        type = "GOLANG";
        name = "40660fec0f8bb30a9452e96abf28ca430faea41a1da5eb8065cbc0e74c8630020708e414f7063885cc8a57def3d4e475672cf5432958b00b04fa75241ff8c7f2";

        userBlockchain = new UserBlockchain(address, port);
    }

    /*
     * (non-Javadoc)
     *
     * @see
     * publickeyblockchain.IPublicKeyBlockchain#registerPublicKey(java.lang.
     * String, java.lang.String)
     */
    @Override
    public boolean registerPublicKey(String user, String publicKey) {
        boolean success = true;
        String[] args = { user, publicKey };

        if (!userBlockchain.getRegistered(user)) {
            blockchain.invoke(type, name, "invoke", args);
        } else {
            success = false;
        }

        return success;
    }

    /*
     * (non-Javadoc)
     *
     * @see publickeyblockchain.IPublicKeyBlockchain#deletePublicKey(java.lang.
     * String, java.lang.String)
     */
    @Override
    public boolean deletePublicKey(String user, String hash) throws JSONException {
        boolean success = true;
        String[] args = { user };

        if (userBlockchain.login(user, hash)) {
            blockchain.invoke(type, name, "delete", args);
        } else {
            success = false;
        }

        return success;
    }

    /*
     * (non-Javadoc)
     *
     * @see
     * publickeyblockchain.IPublicKeyBlockchain#queryPublicKey(java.lang.String)
     */
    @Override
    public String queryPublicKey(String user) {
        String publicKey = null;
        String[] args = { user };

        if (userBlockchain.getRegistered(user)) {
            JSONObject result;

            result = blockchain.query(type, name, "query", args);

            try {
                publicKey = result.getString("OK");

            } catch (JSONException e) {

                e.printStackTrace();
            }

        }

        return publicKey;

    }

}